src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,412 @@
+				// Emoji Shortcode
+				const re = /:[a-z0-9-+_]+$/;
+				const text = value.slice(caretposition < (longest - 1) ? 0 : caretposition - (longest - 1), caretposition);
+				let regexResult = re.exec(text);
+				if (regexResult) {
+					const aregexResult = Object.keys(emojiShortcodes).filter((item) => item.indexOf(regexResult[0]) === 0);
+					if (aregexResult.length === 1 && (regexResult[0].length > 2 || aregexResult[0].length === 3)) {
+						insert = aregexResult[0].slice(regexResult[0].length);
+						output = true;
+					}
+				}
+			}
+			// Convert fractions and mathematical constants to Unicode characters
+			if (!output && fracts) {
+				// Numbers: https://regex101.com/r/7jUaSP/2
+				const re = /[0-9]+(\.[0-9]+)?$/;
+				const prevoustext = value.slice(0, caretposition - 1);
+				let regexResult = re.exec(prevoustext);
+				if (regexResult) {
+					const text = value.slice(0, caretposition);
+					let aregexResult = re.exec(text);
+					if (!aregexResult) {
+						const label = outputLabel(regexResult[0], regexResult[1]);
+						const index = firstDifferenceIndex(label, regexResult[0]);
+						if (index >= 0) {
+							insert = label.slice(index) + (event.keyCode === 13 ? '\n' : insert);
+							deletecount = regexResult[0].length - index + 1;
+							output = true;
+						}
+					}
+				}
+			}
+		}
+		if (output) {
+			const text = value.slice(caretposition - deletecount, caretposition);
+			deleteCaret(target, text);
+			insertCaret(target, insert);
+			console.debug("Autocorrect: “%s” was replaced with “%s”.", text, insert);
+
+			insertedText = insert;
+			deletedText = text;
+
+			lastTarget = target;
+			lastCaretPosition = caretposition - deletecount + insert.length;
+		}
+	}
+}
+
+/**
+ * Undo autocorrect.
+ *
+ * @param {Object} event
+ * @returns {void}
+ */
+function undoAutocorrect(event) {
+	// console.log('keyup', event.key, event.key.length, event.keyCode);
+	if (!(!event.ctrlKey && !event.metaKey && !event.altKey)) {
+		return;
+	}
+	// Backspace
+	if (event.keyCode === 8) {
+		const target = event.target;
+		const caretposition = getCaretPosition(target);
+		if (caretposition) {
+			if (target === lastTarget && caretposition === lastCaretPosition) {
+				event.preventDefault();
+
+				if (insertedText) {
+					deleteCaret(target, insertedText);
+				}
+				if (deletedText) {
+					insertCaret(target, deletedText);
+				}
+				console.debug("Undo autocorrect: “%s” was replaced with “%s”.", insertedText, deletedText);
+			}
+		}
+	}
+
+	lastTarget = null;
+}
+
+/**
+ * Handle response from the autocorrect module.
+ *
+ * @param {Object} message
+ * @param {Object} sender
+ * @returns {void}
+ */
+function handleResponse(message, sender) {
+	if (message.type !== AUTOCORRECT_CONTENT) {
+		return;
+	}
+	autocomplete = message.autocomplete;
+	quotes = message.quotes;
+	fracts = message.fracts;
+	autocorrections = message.autocorrections;
+	longest = message.longest;
+	symbolpatterns = message.symbolpatterns;
```

Comment:
Sry for the delay, I still get the `symbolpatterns = null` error as mentioned above.

I guess this is because the message has not been received here. Setting some breakpoints here or so also shows me nothing is being received.

So maybe:
1. add an explicit check/better error handling when sth. is not received and throw an `Error` with a better description. As that may e.g. also happen when you decline the permission on the options page AFAIK? LIke `throw new Error("emoji autocomplete settings have not been received. Do not autocomplete.")
2. we need to find the cause iof that error

I also tried changing the settings in the options and all other tabs did not receive an update. AFAIK they should, should not they?

I guess the error lies in `AutocorrectHandler.js` somehow.
And indeed, I could fix a camel-case error there, at least – that was pointed out by EsLint. That did not fix this whole problem though… :thinking: 
(I#ve pushed the commit.)

So important: **Please use EsLint** as per our contributing guide.

Anyway, before this can get merged, I need to confirm it works and test that out. So we need to get the cause of this error.

Reply:
1. I updated it to throw your error, but this error should never happen if it is working correctly. The `tabs` permission requested on the options page is only used after the user changes the options, so it should have no effect here.
2. Yes, I agree! I find it very weird that it works for me in mutiple instances of Firefox, but yet it is not working for you. Could there be some Firefox setting or another extension that is interfering with it?


> I also tried changing the settings in the options and all other tabs did not receive an update. AFAIK they should, should not they?

Yes, if you accepted the `tabs` permission.

> And indeed, I could fix a camel-case error there, at least – that was pointed out by EsLint.

If I change the `emojiMart` to `emojimart`, I get this error in the console:
```
Uncaught (in promise) TypeError: emojimart.emojiIndex is undefined
```

> Anyway, before this can get merged, I need to confirm it works and test that out. So we need to get the cause of this error.

Could it possibly be that because of the "submodule error", not all the required dependencies have downloaded? [Here](https://tealdulcet.com/awesome-emoji-picker.zip) is the version I have been testing (note that it also includes full Thunderbird and Chrome support, among other out of scope features).

Reply:
Indeed your ZIP version works. I'll need to investigate that further… 😄